lib/licensed/shell.rb

```diff
@@ -2,27 +2,96 @@
 require "open3"
 
 module Licensed
-  module Shell
-    # Executes a command, returning it's STDOUT on success.  Returns an empty
-    # string on failure
+  class Shell
+    # Executes a command, returning its output on success, and raising an error
+    # on failure.
     def self.execute(cmd, *args)
-      output, _, status = Open3.capture3(cmd, *args)
-      return "" unless status.success?
+      result = new(cmd, *args)
+      output = result.to_s
+      result.raise_on_failure!
       output.strip
     end
 
     # Executes a command and returns a boolean value indicating if the command
     # was succesful
     def self.success?(cmd, *args)
-      _, _, status = Open3.capture3(cmd, *args)
-      status.success?
+      result = new(cmd, *args)
+      result.to_s
+      result.success?
     end
 
     # Returns a boolean indicating whether a CLI tool is available in the
     # current environment
     def self.tool_available?(tool)
-      output, err, status = Open3.capture3("which", tool)
-      status.success? && !output.strip.empty? && err.strip.empty?
+      result = new("which", tool)
+      output = result.to_s
+      result.success? && !output.strip.empty? && result.error_output.strip.empty?
+    end
+
+    class Error < RuntimeError
+      def initialize(cmd, status, stderr)
+        super()
+        @cmd = cmd
+        @exitstatus = status
+        @output = stderr
+      end
+
+      def message
+        output = @output.to_s.strip
+        extra = output.empty?? "" : "\n#{output.gsub(/^/, "    ")}"
+        "command exited with status #{@exitstatus}\n  #{escape_cmd}#{extra}"
+      end
+
+      def escape_cmd
+        @cmd.map do |arg|
+          if arg =~ /[\s'"]/
+            escaped = arg.gsub(/([\\"])/, '\\\\\1')
+            %("#{escaped}")
+          else
+            arg
+          end
+        end.join(" ")
+      end
+    end
+
+    attr_reader :error_output
+
+    def initialize(*cmd)
+      @cmd = cmd
+      @status = nil
+      @error_output = nil
+    end
+
+    def success?
+      @status.success?
+    end
+
+    def exit_code
+      @status.exitstatus
+    end
+
+    # Yields each line as it's being read.
+    def each_line
+      return to_enum(__method__) unless block_given?
+
+      @status = Open3.popen3(*@cmd) do |stdin, stdout, stderr, thread|
```

Comment:
it's a little strange that the command is executed only when calling `#each_line`.  Perhaps it should be executed whenever a "terminal" method is called, e.g. `success?`, `exit_code` etc

Reply:
@jonabc I've defined `each_line` as a “kicker” method here since that's also the method that accepts a block to be yielded in line-by-line streaming mode. If another method kicked off the process, how would you envision `each_line` accepting a block?

Thanks for your comments! I've addressed other concerns in the meantime.